internal/controllers/reference/reference.go

```diff
@@ -115,6 +155,64 @@ func removeOutdatedReferencesToSecret(
 	return nil
 }
 
+// removeOutdatedReferencesToConfigMap removes outdated reference records to configMaps in reference indexer.
+// ConfigMaps that are referred by referrer are passed in referredConfigMapNameMap parameter.
+// If a ConfigMap is not referenced by any other object after deleting outdated reference records,
+// and it does not have label "konghq.com/ca-cert:true", it is not possible to be used in Kong gateway config
+// and should be removed from the object cache inside KongClient.
+func removeOutdatedReferencesToConfigMap(
+	ctx context.Context,
+	indexers CacheIndexers, c client.Client, dataplaneClient controllers.DataPlaneClient,
+	referrer client.Object, referredConfigMapNameMap map[k8stypes.NamespacedName]struct{},
+) error {
+	referents, err := indexers.ListReferredObjects(referrer)
+	if err != nil {
+		return err
+	}
+	for _, obj := range referents {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		// delete the reference record if the secret is not referred by the service.
+		if gvk.Group == corev1.GroupName && gvk.Version == VersionV1 && gvk.Kind == KindConfigMap {
+			namespacedName := k8stypes.NamespacedName{
+				Namespace: obj.GetNamespace(),
+				Name:      obj.GetName(),
+			}
+
+			// if the configmap is still referenced, no operations are taken so continue here.
+			if _, ok := referredConfigMapNameMap[namespacedName]; ok {
+				continue
+			}
+
+			if err := indexers.DeleteObjectReference(referrer, obj); err != nil {
+				return err
+			}
+			// remove the configMap in object cache if it is not referred and does not have label "konghq.com/ca-cert:true".
+			// Do this check and delete when the reference count may be reduced by 1.
+
+			// retrieve the configMap in k8s and check it has the label.
+			configMap := &corev1.ConfigMap{}
+			getErr := c.Get(ctx, namespacedName, configMap)
+			// if the configMap exists in k8s and has the label, we should not delete it in object cache.
+			if getErr == nil {
+				if configMap.Labels != nil && configMap.Labels[CACertLabelKey] == "true" {
+					continue
+				}
+			} else {
+				// if the configMap does not exist in k8s, we ignore the error and continue the check and delete operation.
+				// for other errors, we return the error and stop the operation.
+				if !apierrors.IsNotFound(getErr) {
+					return err
+				}
+			}
```

Comment:
```suggestion
			// if the configMap exists in k8s and has the label, we should not delete it in object cache.
			if getErr != nil {
			        // if the configMap does not exist in k8s, we ignore the error and continue the check and delete operation.
				// for other errors, we return the error and stop the operation.
				if !apierrors.IsNotFound(getErr) {
					return err
				}
			}
			if configMap.Labels != nil && configMap.Labels[CACertLabelKey] == "true" {
					continue
			}
```
Will it be the same, won't it? For me easier to follow

Reply:
Practically yes, let's go this way 👍